Sources/System/Internals/WindowsSyscallAdapters.swift

```diff
@@ -115,11 +114,144 @@ internal func pwrite(
   var nNumberOfBytesWritten: DWORD = 0
   if !WriteFile(hFile, buf, DWORD(nbyte), &nNumberOfBytesWritten,
                 &ovlOverlapped) {
-    let _ = GetLastError()
-    // TODO(compnerd) map windows error to errno
+    ucrt._set_errno(mapWindowsErrorToErrno(GetLastError()))
     return Int(-1)
   }
   return Int(nNumberOfBytesWritten)
 }
 
+@inline(__always)
+internal func pipe(_ fds: UnsafeMutablePointer<Int32>) -> CInt {
+  return _pipe(fds, 4096, _O_BINARY | _O_NOINHERIT);
+}
+
+@inline(__always)
+internal func ftruncate(_ fd: Int32, _ length: off_t) -> Int32 {
```

Comment:
The entire C library on Windows is in some sense "emulated", since the C library is not the native platform API there. I think this likely explains the difference in approach from Saleem in the past. And yes, the UCRT doesn't have `ftruncate()` or any equivalent (that I'm aware of), but the Win32 API totally does all the things we need it to do to make that work, and it isn't terribly complicated to implement.

Note: I'm _not_ arguing for one moment that `swift-system` should become a POSIX emulation layer. Just observing that in practice, non-POSIX _largely_ means Windows at this point, and if anyone is using `FileDescriptor` for anything then they'll end up writing things that aren't portable to Windows unless we make sure that we have equivalent functionality. I don't think it makes sense to push yet more `#if os(Windows)` stuff up the stack.

(I'm also playing Devil's advocate here.)

Reply:
Anything importing SystemPackage is expected to end up having to deal with `#if os(Foo)` conditions, anyway. 

SystemPackage is not designed to provide cross-platform API.

Reply:
Sure. We spoke about this on Slack… fundamentally the issue here is that from a Windows developer's standpoint, a (kind of) subset of the POSIX file descriptor API does exist on Windows and is useful for porting tools written for UNIX. It isn't all that useful for Windows programming more generally — Windows developers would generally drop down to the Win32 layer rather than using `_open()` et al — everyone knows that it's all emulated anyway.

Basically, I don't think it's worth _not_ supporting things that are trivially easy to implement, like `pipe()` and `ftruncate()`. I also don't think it should evolve into a Cygwin like layer, because that's _way_ too far.

I thought we'd agreed that this was the right thing for Windows; maybe I misunderstood the conversation we had?